File: ios/Sources/RynBridge/ModuleRegistry.swift
import Foundation

public final class ModuleRegistry: @unchecked Sendable {
    private let lock = NSLock()
    private var modules: [String: BridgeModule] = [:]

    public init() {}

    public func register(_ module: BridgeModule) {
        lock.lock()
        defer { lock.unlock() }
        modules[module.name] = module
    }

    public func getAction(module moduleName: String, action actionName: String) throws -> ActionHandler {
        lock.lock()
        defer { lock.unlock() }

        guard let module = modules[moduleName] else {
            throw RynBridgeError(code: .moduleNotFound, message: "Module '\(moduleName)' not found")
        }
        guard let handler = module.actions[actionName] else {
            throw RynBridgeError(code: .actionNotFound, message: "Action '\(actionName)' not found in module '\(moduleName)'")
        }
        return handler
    }

    public func hasModule(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return modules[name] != nil
    }
}

File: android/core/src/main/kotlin/com/devryner/rynbridge/core/ModuleRegistry.kt
package com.devryner.rynbridge.core

class ModuleRegistry {

    private val lock = Any()
    private val modules = mutableMapOf<String, BridgeModule>()

    fun register(module: BridgeModule) {
        synchronized(lock) {
            modules[module.name] = module
        }
    }

    fun getAction(module: String, action: String): ActionHandler {
        synchronized(lock) {
            val mod = modules[module]
                ?: throw RynBridgeError(
                    code = ErrorCode.MODULE_NOT_FOUND,
                    message = "Module '$module' not found"
                )
            return mod.actions[action]
                ?: throw RynBridgeError(
                    code = ErrorCode.ACTION_NOT_FOUND,
                    message = "Action '$action' not found in module '$module'"
                )
        }
    }

    fun hasModule(name: String): Boolean {
        synchronized(lock) {
            return modules.containsKey(name)
        }
    }
}
